### ISSUE TITLE
[spotify] Bridge deviceName & deviceId channels are the same

### ISSUE THREAD

**Issue by Grimthorr** [openhab/openhab2-addons#5104]
## Expected Behavior
Within the Spotify binding: the `deviceName` channel should subscribe to the state of the currently active device *name*, whereas the `deviceId` channel should be the device *ID*.

## Current Behavior
Both `deviceName` and `deviceId` channels are receiving the state of the currently active device *ID*.

## Possible Solution
Change line 443 of `SpotifyBridgeHandler.java` to use `device.getName()` instead of `device.getId()`.

https://github.com/openhab/openhab2-addons/blob/89821ad36c08de584142a6b808a8b66755dac6e6/addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/handler/SpotifyBridgeHandler.java#L440-L444

Pull request incoming...

**Comment by Hilbrand** [comment]
What is the problem? The deviceName channel uses dynamic StateOption. Where Id is the key and name the value. Therefore to set the channel it must be set with the id. I've seen this doesn't work always correctly, for example when all devices are off line it seems to keep the last value, the id. So I don't think the solution is to set the name.

**Comment by Grimthorr** [comment]
Thanks @Hilbrand. When reading the value of the `deviceName` channel (via a linked `String` item), it was giving me the device ID and not the name, so I took a brief glance at the source and assumed the above was the cause. When I made the PR change, the item I have attached to the `deviceName` channel was receiving the name value as I'd hoped.

Am I doing something wrong here? How do I match the provided device ID to the actual name of the device as presented by the Spotify Web API? It just seemed strange that the `deviceName` channel is giving me the same value as the `deviceId` channel.

**Comment by Hilbrand** [comment]
Reading your comment I guessing your trying to read the name in a rule? It seems that when reading the state in a rule it can only return the used key (which is the id here) and not the value. I think this is a limitation of openHAB currently. I'n not sure what the best solution is, possible add a new channel so we have one that show the name, and one that can be used to show the list of names.

**Comment by Grimthorr** [comment]
Ah that's correct; I was reading the value through any of the following methods:

- read the value within a rule
- print the value in HABPanel
- read the value through the console
- logged under `events.log`

In all of these scenarios, it just returns the selected key, not the value. I see it's supposed to be a `Selection` element, and openHAB doesn't have a matching Item type for this? So, the same applies for the `playlist` channel, albeit this uses a slightly different implementation.

Maybe the current `deviceName` channel would be better named `devices` which acts as a device list and then the `deviceName` channel can be used for the currently active device name as a string? I see benefits for a similar `playlistName` channel too. I guess I'll leave this up to you though as I didn't realise how this worked!

Can you propose a workaround to get the value from the key for both `deviceName` and `playlist`? How do you programmatically get the key>value pairs for these? Or, for that matter, how does a Selection element do it?

Thanks for your assistance & patience with this!

**Comment by Grimthorr** [comment]
On a similar note, the Web API seems to provide playlist URIs inconsistently.

When the `playlist` channel is populated, the keys are in the format: `spotify:playlist:<id>`, but the currently active playlist is returned in a different format: `spotify:user:<owner>:playlist:<id>`, so the selection is mismatched.

I can raise a separate issue for this if you can replicate?

**Comment by Grimthorr** [comment]
I've found a way to access the underlying values of dynamic state items within rules. For anyone else finding this, here's a quick example:

The `deviceName` and `playlist` channels use the [`SpotifyDynamicStateDescriptionProvider`](https://github.com/openhab/openhab2-addons/blob/be3f3a09db76ab7af9c61cdf63e55cb2096c68fe/addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/handler/SpotifyDynamicStateDescriptionProvider.java#L39) to hold the actual lists and therefore provide a `StateDescription` to match the key-value pairs. This means that you can call `getStateDescription.getOptions` on an item linked to the channel to pull the list of pairs, for example:

```
rule "Spotify device name"
when
    Item Spotify_Device_Name changed
then
    Spotify_Device_Name.getStateDescription.getOptions.forEach[option |
        logInfo("Spotify", "Device {}: {}", option.getValue, option.getLabel)
        if( Spotify_Device_Name.state == option.getValue ) {
            logInfo("Spotify", "Current device: {}", option.getLabel)
            // Do stuff
        }
    ]
end
```

In the above example, the `Spotify_Device_Name` item is linked to a Spotify bridge `deviceName` channel. The same could be applied to a linked `playlist` item.

**Comment by bulletprooffool** [comment]
> I've found a way to access the underlying values of dynamic state items within rules. For anyone else finding this, here's a quick example:
> 
> The `deviceName` and `playlist` channels use the [`SpotifyDynamicStateDescriptionProvider`](https://github.com/openhab/openhab2-addons/blob/be3f3a09db76ab7af9c61cdf63e55cb2096c68fe/addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/handler/SpotifyDynamicStateDescriptionProvider.java#L39) to hold the actual lists and therefore provide a `StateDescription` to match the key-value pairs. This means that you can call `getStateDescription.getOptions` on an item linked to the channel to pull the list of pairs, for example:
> 
> ```
> rule "Spotify device name"
> when
>     Item Spotify_Device_Name changed
> then
>     Spotify_Device_Name.getStateDescription.getOptions.forEach[option |
>         logInfo("Spotify", "Device {}: {}", option.getValue, option.getLabel)
>         if( Spotify_Device_Name.state == option.getValue ) {
>             logInfo("Spotify", "Current device: {}", option.getLabel)
>             // Do stuff
>         }
>     ]
> end
> ```
> 
> In the above example, the `Spotify_Device_Name` item is linked to a Spotify bridge `deviceName` channel. The same could be applied to a linked `playlist` item.

Thanks for this.
I am desperately trying to find a way to idetify the current playing device by NAME rather than id, since the id on my Echo keeps changing for some reason.

I have tried using the code snippet you posted above, but it is not even logging the items. I know I am entering your loop, since I added a log right after the 'then' and before your enumeration loop - but nothing.
Am I missing something to make this work?

Thanks for your help.

**Comment by Hilbrand** [comment]
Use the latest version of the bindings.  I've changed these channels and how devices are identified. See https://community.openhab.org/t/spotify-connect-binding-available/66809

**Comment by Grimthorr** [comment]
@bulletprooffool Below is the final rule that I'm using to extract the device name from the list of devices:

```
rule "Spotify device name"
when
    Item Spotify_Current_Devices changed
then
    Spotify_Current_Devices.getStateDescription.getOptions.forEach[option |
        if( Spotify_Current_Devices.state == option.getValue ) {
            logInfo("Spotify", "Current device name: {}", option.getLabel)
            postUpdate("Spotify_Current_Device_Name", option.getLabel)
        }
    ]
end
```

With the associated items:

```
String Spotify_Current_Devices {channel="spotify:player:2687b1d5:deviceName"}
String Spotify_Current_Device_Name
```

This works for me using version `market:binding-4326451 - 1.0` of the Spotify binding from the Eclipse Marketplace.

@Hilbrand Does the latest version incorporate commits from #5434? Thanks for these changes, it will definitely solve the initial confusion raised here. These 2 issues should be linked for visibility.

How do we force openHAB to update the binding? And when are new bindings like this made available in the openHAB addons list (without using Eclipse Marketplace)?

**Comment by bulletprooffool** [comment]
Awesome - thanks, this did the trick

**Comment by Hilbrand** [comment]
@Grimthorr Yes the latest version does incorporate the commits.

You can force openHAB to update by uninstalling and than re-installing the binding. For update information see the first post in this thread on the forum: https://community.openhab.org/t/spotify-connect-binding-available/66809

The Spotify binding is available in 2.5 snapshot (although I don't recommend updating to snapshot only for this binding. The market place binding does the same). It will also be available once a release of openHAB is done.
